### ISSUE TITLE
infinite syncing loop

### ISSUE THREAD

**Issue by stratospark** [electric-sql/electric#1552]
I'm getting this issue where the frontend repeatedly tries to sync data after getting 409 errors. It seems as if the offset being requested is incorrect.

I've created a simple reproduction where modifying two tables within a transaction creates this behavior.

https://github.com/stratospark/electricsql-broken-sync-repro

**Comment by alco** [comment]
Hey @stratospark :wave: 

Brilliant work on the report and the reproduction repo! One suggestion: use the `electricsql/electric` Docker Hub image if you want to run the latest published release or `electricsql/electric:canary` to run the latest build from the `main` branch. `electricsql/electric-next` is no longer updated.

I can confirm that the behaviour you've described is reproducible using the latest `canary` image of `electricsql/electric`.

**Comment by kevin-dp** [comment]
![Screenshot 2024-08-22 at 10 27 19](https://github.com/user-attachments/assets/6812f51b-2847-44c5-b03b-ba3bdd34f9ca)

Looking at the network activity i notice that upon receiving a 409, the response contains a `location` header for the new location we need to hit and also the last offset and the shape ID. Looking and subsequent requests, it seems that the client is hitting the new location but also still retrying the old one.

I believe it is due to this part of the code in the JS client:
```ts
let response!: Response
try {
  const maybeResponse = await this.fetchWithBackoff(fetchUrl)
  if (maybeResponse) response = maybeResponse
  else break
} catch (e) {
  if (!(e instanceof FetchError)) throw e // should never happen
  if (e.status == 409) {
    // Upon receiving a 409, we should start from scratch
    // with the newly provided shape ID
    const newShapeId = e.headers[`x-electric-shape-id`]
    this.reset(newShapeId)
    this.publish(e.json as Message[])
    continue
  } else if (e.status >= 400 && e.status < 500) {
    // Notify subscribers
    this.sendErrorToUpToDateSubscribers(e)
    this.sendErrorToSubscribers(e)
    // 400 errors are not actionable without additional user input, so we're thr
    throw e
  }
}
const { headers, status } = response
const shapeId = headers.get(`X-Electric-Shape-Id`)
if (shapeId) {
  this.shapeId = shapeId
}
const lastOffset = headers.get(`X-Electric-Chunk-Last-Offset`)
if (lastOffset) {
  this.lastOffset = lastOffset as Offset
}
```

When hitting a 409 we do `this.reset(newShapeId)` which causes the JS client to reset the offset to -1 and thus send requests to the new location. But then we do `continue` which causes the JS client to override the `shapeId` and the `lastOffset` with the one from the response headers (which are the old ones).

**Comment by kevin-dp** [comment]
Seems like it is a loop caused by the server:
- We get a 409 telling us to refetch with offset -1
- Then we refetch with offset -1 and server gives us back the old offset for which we got 409
- stuck in this loop forever

**Comment by KyleAMathews** [comment]
Fix is released now in 0.3.5. Do a docker pull and you'll be back in business!
